### PR TITLE
Add recycle bin

### PR DESCRIPTION
**Pull request by alichaddad**
Addresses:
- #1555 
- #1551 
- #1535

### CODE REVIEW

File: apps/cloudbroker/cloudbrokerlib/authenticator.py [hunk 1]
```diff
@@ -191,6 +191,8 @@ def checkCloudspaceStatus(self, requiredaccessrights, cloudspace):
         """
         if cloudspace.status == 'DESTROYED':
```
[the remark this answers is not in the document]
Comment (chrisvdg) [reply apps/cloudbroker/cloudbrokerlib/authenticator.py]: Also in the line underneath this comment

File: apps/cbportal/base/CBGrid/.macros/page/cbbin/3_cbbin.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+def main(j, args, params, tags, tasklet):
+    import cgi
+    page = args.page
+    page.addJS('/system/.files/js/action.js', header=False)
+    modifier = j.html.getPageModifierGridDataTables(page)
+    filters = {'status': 'DELETED'}
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/.macros/page/cbbin/3_cbbin.py]: use enums

File: apps/cbportal/base/CBGrid/.macros/page/cbspaces/3_cbspaces.py [hunk 3]
```diff
@@ -2,7 +2,7 @@ def main(j, args, params, tags, tasklet):
     page = args.page
     modifier = j.html.getPageModifierGridDataTables(page)
 
-    filters = {'status': {'$ne': 'DESTROYED'}}
+    filters = {'status': {'$nin': ['DESTROYED', 'DELETED']}}
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/.macros/page/cbspaces/3_cbspaces.py]: please use the enums

File: apps/cbportal/base/CBGrid/.macros/page/cbspacesusers/3_cbspaces.py [hunk 4]
```diff
@@ -3,7 +3,7 @@ def main(j, args, params, tags, tasklet):
     page = args.page
     modifier = j.html.getPageModifierGridDataTables(page)
 
-    filters = dict()
+    filters = {'status': {'$nin': ['DESTROYED', 'DELETED']}}
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/.macros/page/cbspacesusers/3_cbspaces.py]: enums

File: apps/cbportal/base/CBGrid/.macros/page/diskbin/3_diskbin.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
+def main(j, args, params, tags, tasklet):
+    import cgi
+    page = args.page
+    modifier = j.html.getPageModifierGridDataTables(page)
+    disktype = args.getTag('type')
+    filters = {'status': 'TOBEDELETED', 'type': disktype}
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/.macros/page/diskbin/3_diskbin.py]: enums

File: apps/cbportal/base/CBGrid/.macros/page/vmachines/3_vmachines.py [hunk 6]
```diff
@@ -7,7 +7,7 @@ def main(j, args, params, tags, tasklet):
     cloudspaceId = args.getTag("cloudspaceid")
     imageid = args.getTag('imageid')
     gid = args.getTag('gid')
-    filters = {'status': {'$ne': 'DESTROYED'}}
+    filters = {'status': {'$nin': ['DESTROYED', 'DELETED']}}
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/.macros/page/vmachines/3_vmachines.py]: enums

File: apps/cbportal/base/CBGrid/.macros/page/vmbin/3_vmbin.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
+def main(j, args, params, tags, tasklet):
+    import cgi
+    page = args.page
+    modifier = j.html.getPageModifierGridDataTables(page)
+    filters = {'status': 'DELETED'}
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/.macros/page/vmbin/3_vmbin.py]: enums

File: apps/cbportal/base/CBGrid/.macros/wiki/binids/3_binids.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
+def main(j, args, params, tags, tasklet):
+    params.result = (args.doc, args.doc)
+
+    def _get_data(data):
+        return [val['id'] for val in data]
+
+    ccl = j.clients.osis.getNamespace('cloudbroker')
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/.macros/wiki/binids/3_binids.py]: enums

File: apps/cbportal/base/CBGrid/Cloud Spaces/Cloud Space.wiki [hunk 9]
```diff
@@ -72,14 +74,21 @@ h2. Cloud Space: ${name}
 
 {% endif %}
 
-- display: Delete Cloud Space
+- display: Delete cloud Space
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/Cloud Spaces/Cloud Space.wiki]: Titles are uppercased why did you lower this? PLease uppercase again make sure to refer to it correctly from any actions

File: apps/cbportal/base/CBGrid/Cloud Spaces/Cloud Space.wiki [hunk 10]
```diff
@@ -141,6 +150,26 @@ h2. Cloud Space: ${name}
...
+
+{{actions:
+- display: Restore Cloud Space
+  action: /restmachine/cloudbroker/cloudspace/restore
+  data:
+   cloudspaceId: $$id
+  input:
+  - reason
+
+- display: Permenant delete
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/Cloud Spaces/Cloud Space.wiki]: Capital D please

File: apps/cbportal/base/CBGrid/Recycle Bin.wiki [hunk 11]
```diff
@@ -0,0 +1,167 @@
+@usedefault
+
+@@title Recycle bin
+
+{{binids}}
+
+h1. Recycle Bin
+
+{{actions:
+- display: Destroy Cloudspace
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/Recycle Bin.wiki]: Cloud Space <- is used everywhere else...

File: apps/cbportal/base/CBGrid/Recycle Bin.wiki [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+  action: /restmachine/cloudbroker/cloudspace/destroy
+  hide: true
+  refresh: true
+  input:
+  - label: CloudSpace name
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/Recycle Bin.wiki]: Cloud Space 2 words

File: apps/cbportal/base/CBGrid/Recycle Bin.wiki [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+    type: text
+    name: name
+    required: true
+  - label: Reason
+    type: text
+    name: reason
+  data:
+    permanently: True
+- display: Restore Cloudspace
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/Recycle Bin.wiki]: Cloud Space

File: apps/cbportal/base/CBGrid/Recycle Bin.wiki [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+- display: Restore Machine
+  action: /restmachine/cloudbroker/machine/restore
+  hide: true
+  refresh: true
+  input:
+  - label: Reason
+    type: text
+    name: reason
+
+- display: Destroy Cloudspaces
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/Recycle Bin.wiki]: Cloud Spaces

File: apps/cbportal/base/CBGrid/Recycle Bin.wiki [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+    type: text
+    name: reason
+}}
+
+
+h2. Deleted Cloud Spaces
+{{datatables_use}}
+
+{{actions:
+- display: Destroy all Cloudspaces
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/Recycle Bin.wiki]: Destroy All Cloud Spaces

File: apps/cbportal/base/cloudbroker__cloudspace/methodclass/cloudbroker_cloudspace.py [hunk 16]
```diff
@@ -62,40 +65,53 @@ def _destroy(self, cloudspace, reason, ctx):
...
             raise
 
-        # delete routeros
-        ctx.events.sendMessage(title, 'Deleting Virtual Firewall')
-        self._destroyVFW(cloudspace['gid'], cloudspace['id'])
-        cloudspace = self.models.cloudspace.get(cloudspace['id'])
-        cloudspace.status = resourcestatus.Cloudspace.DESTROYED
-        self.cb.cloudspace.release_resources(cloudspace)
+        if not permanently:
+            fwid = '%s_%s' % (cloudspace['gid'], cloudspace['networkId'])
```
Comment (grimpy) [inline apps/cbportal/base/cloudbroker__cloudspace/methodclass/cloudbroker_cloudspace.py]: would be nice to send event here indication you ar estopping VFW like on line 84

File: apps/cbportal/base/cloudbroker__cloudspace/methodclass/cloudbroker_cloudspace.py [hunk 17]
```diff
@@ -279,6 +295,8 @@ def startVFW(self, cloudspaceId, **kwargs):
         """
         cloudspaceId = self._getCloudSpace(cloudspaceId)['id']
         cloudspace = self.models.cloudspace.get(cloudspaceId)
+        if cloudspace.status == resourcestatus.Cloudspace.DELETED:
+            raise exceptions.BadRequest('Selected cloudspace is deleted')
```
Comment (grimpy) [inline apps/cbportal/base/cloudbroker__cloudspace/methodclass/cloudbroker_cloudspace.py]: Cloud Space

File: apps/cloudbroker/base/cloudapi__disks/methodclass/cloudapi_disks.py [hunk 18]
```diff
@@ -246,3 +252,21 @@ def resize(self, diskId, size, **kwargs):
         if not res:
             raise exceptions.Accepted(False)
         return True
+
+    @authenticator.auth(acl={'cloudspace': set('X')})
+    def restore(self, diskId, reason, **kwargs):
+        disk = self.models.disk.searchOne({'id': diskId})
+        if not disk or disk['status'] == resourcestatus.Disk.DESTROYED:
+            raise exceptions.NotFound("Couldn't find disk with id: %s" % diskId)
+        if disk['status'] != resourcestatus.Disk.TOBEDELETED:
```
Comment (grimpy) [inline apps/cloudbroker/base/cloudapi__disks/methodclass/cloudapi_disks.py]: werent we gonna change this enum name?

Comment (alichaddad) [reply apps/cloudbroker/base/cloudapi__disks/methodclass/cloudapi_disks.py]: I changed the behavior now. `DELETED` state is for deleted attached disks and `TOBEDELETED` for unattached disks. I know this is not optimal and i am open to suggestions.

File: apps/cbportal/base/CBGrid/Images/Image.wiki [hunk 19]
```diff
@@ -4,6 +4,9 @@
 {% if id %}
 h3. Image: ${name}
 
+{% if status != "DESTROYED" %}
+{% if status != "DELETED" %}
```
Comment (grimpy) [inline apps/cbportal/base/CBGrid/Images/Image.wiki]: two ifs and 2 endifs  you can just do an `and` or `in` right?

File: apps/cloudbroker/base/cloudapi__disks/methodclass/cloudapi_disks.py [hunk 20]
```diff
@@ -267,6 +272,7 @@ def restore(self, diskId, reason, **kwargs):
     def deleteDisks(self, diskIds, reason, **kwargs):
         for diskId in diskIds:
             self.delete(diskId, False, True, reason=reason)
+            return True
```
Comment (grimpy) [inline apps/cloudbroker/base/cloudapi__disks/methodclass/cloudapi_disks.py]: this return True appears to be in the middle of a loop?

File: apps/cloudbroker/base/cloudapi__machines/methodclass/cloudapi_machines.py [hunk 21]
```diff
@@ -113,12 +114,19 @@ def start(self, machineId, diskId=None, **kwargs):
...
                 raise exceptions.BadRequest("diskId is not of type CD-ROM")
+            if disk.status in resourcestatus.Image.INVALID_STATES:
+                raise exceptions.BadRequest("Rescue disk with id {} is not available".format(diskId))
             cloudspace = self.models.cloudspace.get(machine.cloudspaceId)
             if disk.accountId and disk.accountId != cloudspace.accountId:
                 raise exceptions.BadRequest("Rescue disk with id {} belongs to another account".format(diskId))
             volume = j.apps.cloudapi.disks.getStorageVolume(disk, provider, node)
             node.extra['volumes'].append(volume)
             node.extra['bootdev'] = 'cdrom'
+            tags.tagSet('cdrom', disk.id)
```
Comment (grimpy) [inline apps/cloudbroker/base/cloudapi__machines/methodclass/cloudapi_machines.py]: tags need to be string I beleive sthi will crash you need to do str(disk.id)

File: apps/cbportal/base/CBGrid/Images/Image.wiki (hunk 19 above)
Comment (alichaddad) [reply apps/cbportal/base/CBGrid/Images/Image.wiki]: No this is needed.

File: apps/cloudbroker/base/cloudapi__machines/methodclass/cloudapi_machines.py (hunk 21 above)
Comment (alichaddad) [reply apps/cloudbroker/base/cloudapi__machines/methodclass/cloudapi_machines.py]: No it doesn't have to be

**Pull request merged**